### PR TITLE
Fix for #67

### LINKED ISSUE #67: Copying tl_news with tl_content results in double translations for the copied tl_content record
Hello,

I think this is related to https://github.com/terminal42/contao-DC_Multilingual/issues/14.

Steps to reproduce:

1. Create a tl_news (DC_Multilingual)
2. Create a language record for the news
3. Create a tl_content (DC_Multilingual) which is a child of the news created before
4. Create a language record for the content element
5. Copy the news

Result:

The `langPid` column is not updated and is still pointing to the *copied, original* `tl_news` record which is causing the translation to appear doubled in the frontend.

It seems to be the following line:

https://github.com/terminal42/contao-DC_Multilingual/blob/master/src/Driver.php#L682

Thanks in advance for digging into it!

Bye Defcon0

### PR DESCRIPTION
**Pull request by Defcon0**
OK, I think I solved the issue :-)

In the first place I wanted to do it similar to the existing solution, i.e., running the `parent::copyChilds()` and then do the langPid adjustments. In fact this wasn't possible because of the following reason:

Consider a `tl_news` instance with 2 `tl_content` instances inside it. When copying the news, Contao's default implementation (`copyChilds()`) copies all children with the `langPid` column being wrong afterwards. Now if you want to correct that, you simply can't because you can't tell which **new** translation belongs to which **new** record. You only have the relation from **old** original to **old** translation record, but not from **old** original to **new** original. This leads to you being unable to tell what to put inside the `langPid` fields of the copied records.

My solution was to generate a mapping array to contain all necessary old vs. new id mapping information. Afterwards I apply this information to correct the `langPid` fields. This has also the advantage that we now use the recursion correctly.

This PR also solves another issue. When copying a `tl_news` instance which already contains one or more translations, the `pid` column was 0 after the `copy()` call has been processed. This leads to the translation being deleted by Contao if the normal record (the non-translation one) is submitted (saved). Same for the field `ptable` for dynamic parent tables.

It's been a bit complicated, but my tests have all been successful. The test cases have been:

- only `tl_news` is translatable -> copied correctly
- `tl_news` and `tl_content` are translatable -> copied correctly
- `tl_news_archive`, `tl_news` and `tl_content` are translatable -> copied correctly

Could you please test it and merge if ok?

Thanks and have a nice day!

### CODE REVIEW

**Comment by Toflar** [comment]
Hey @dmolineus, as you've been contributing a lot to this lately, do you mind reviewing this piece of code?
I haven't used this DC in quite a while and I have no project where I can verify at all.

**Comment by dmolineus** [comment]
@Toflar I'll take a look at it.

**Comment by Defcon0** [comment]
Thanks in advance, @dmolineus :-)

File: src/Driver.php [hunk 1]
```diff
@@ -618,7 +618,7 @@ public function copy($blnDoNotRedirect = false)
                     if (!$GLOBALS['TL_DCA'][$this->strTable]['fields'][$k]['eval']['translatableFor']) {
                         if (isset($GLOBALS['TL_DCA'][$this->strTable]['fields'][$k]['default'])) {
                             $set[$k] = $GLOBALS['TL_DCA'][$this->strTable]['fields'][$k]['default'];
-                        } else {
+                        } elseif ($k !== 'pid' && $k !== 'ptable') { // skip ptable and pid from being unset since being set to 0 leads to contao deleting the records
```
Comment (dmolineus) [inline src/Driver.php:621]: This fix seems not to fix all issues.

1. A `ptable` might be dynamic and differ from the origin location, so https://github.com/contao/contao/blob/f7ca2e7374e9070cd350d2330f65aa84eac283b1/core-bundle/src/Resources/contao/drivers/DC_Table.php#L946 should be used as well
2. The `pid` might change, Contao handles it in the `getNewsPosition` method: https://github.com/contao/contao/blob/f7ca2e7374e9070cd350d2330f65aa84eac283b1/core-bundle/src/Resources/contao/drivers/DC_Table.php#L1205-L1361

**Review by dmolineus — CHANGES_REQUESTED** [review]

**Comment by dmolineus** [comment]
The proposed fix has a huge limitation: It only works if the data container which record is copied, has the `Multilingual` driver. What if the parent uses the `Table` driver?

Following your example: I have changes `tl_news` to be translatable and also manage content elements for each translated news. If I do a deep copy of the news archive now, this fix will never be called. I guess we need to use the `oncopy_callback` callback to find a general solution.

**Comment by Defcon0** [comment]
Mhm, you're right. I didn't consider that :-( In my case, tl_news and tl_content have been translatable.

So you'd suggest to add an oncopy_callback to all DCAs regardless of whether it uses the Multilingual driver or not and therein scan for the child entities and do the copying?

**Comment by dmolineus** [comment]
Exactly. The difficulty will be to generate the id mapping. Following the order of the record of the source it should be possible.

I do something similar in contao-bootstrap, where the parent relation get fixed independently from with data record is copied:

1. Register the `oncopy_callback` to all data containers: https://github.com/contao-bootstrap/grid/blob/master/src/Listener/Hook/RegisterFixContentParentRelationsFixerListener.php
2. The callback implementation: https://github.com/contao-bootstrap/grid/blob/master/src/Listener/Dca/ParentFixContentParentRelationsListener.php

Maybe this gives you an idea how to start.

**Comment by aschempp** [comment]
As discussed with @Defcon0, I'm closing this PR as it does not fully fix the issue. If someone can come up with a working solution, feel free to open a new PR 😊

**Pull request closed without merging**